File: Sources/WInstallerCore/CommandRunner.swift
import Foundation

/// The result of running a single command.
///
/// Command output is treated as untrusted input (see `TERMINAL_AUTOMATION.md`).
/// Callers must parse it defensively and never evaluate it as code.
public struct CommandResult: Sendable, Equatable {
    public var executable: String
    public var arguments: [String]
    public var standardOutput: Data
    public var standardError: Data
    public var exitCode: Int32
    public var startedAt: Date
    public var finishedAt: Date

    public init(
        executable: String,
        arguments: [String],
        standardOutput: Data,
        standardError: Data,
        exitCode: Int32,
        startedAt: Date,
        finishedAt: Date
    ) {
        self.executable = executable
        self.arguments = arguments
        self.standardOutput = standardOutput
        self.standardError = standardError
        self.exitCode = exitCode
        self.startedAt = startedAt
        self.finishedAt = finishedAt
    }

    public var succeeded: Bool { exitCode == 0 }

    public var standardOutputString: String {
        String(decoding: standardOutput, as: UTF8.self)
    }

    public var standardErrorString: String {
        String(decoding: standardError, as: UTF8.self)
    }
}

/// Timeout classes recommended by `TERMINAL_AUTOMATION.md`.
public enum CommandTimeout: Sendable {
    /// Fast metadata command (10s).
    case metadata
    /// Mount / unmount (60s).
    case mount
    /// Format (120s).
    case format
    /// Eject (60s).
    case eject
    /// Long-running work watched by a progress watchdog rather than a hard timeout.
    case watched
    /// Explicit override in seconds.
    case seconds(TimeInterval)

    public var interval: TimeInterval? {
        switch self {
        case .metadata: 10
        case .mount, .eject: 60
        case .format: 120
        case .watched: nil
        case .seconds(let value): value
        }
    }
}

/// A minimal thread-safe flag used to signal a timeout across the watchdog task.
final class BoolBox: @unchecked Sendable {
    private let lock = NSLock()
    private var _value = false
    var value: Bool {
        get { lock.withLock { _value } }
        set { lock.withLock { _value = newValue } }
    }
}

public enum CommandRunnerError: Error, Equatable, Sendable {
    case launchFailed(String)
    case timedOut(executable: String, seconds: TimeInterval)
    case cancelled
    case nonZeroExit(executable: String, exitCode: Int32, standardError: String)
}

/// Abstraction over command execution so the whole pipeline stays testable
/// without touching a real disk. Commands are always represented as an
/// executable plus an argument array — never a shell-interpolated string.
public protocol CommandRunning: Sendable {
    func run(_ command: PlannedCommand, timeout: CommandTimeout) async throws -> CommandResult
}

public extension CommandRunning {
    func run(_ command: PlannedCommand) async throws -> CommandResult {
        try await run(command, timeout: .metadata)
    }
}

/// Executes commands for real with Foundation `Process`. Captures stdout/stderr
/// and exit status, supports Task cancellation, and enforces the timeout classes
/// above.
///
/// Output is redirected to temporary files rather than pipes so a large stream
/// (e.g. `rsync --info=progress2`) cannot deadlock against a full pipe buffer.
public final class ProcessCommandRunner: CommandRunning {
    public init() {}

    public func run(_ command: PlannedCommand, timeout: CommandTimeout) async throws -> CommandResult {
        try Task.checkCancellation()

        let tmp = FileManager.default.temporaryDirectory
        let outURL = tmp.appendingPathComponent("winstaller-\(UUID().uuidString).out")
        let errURL = tmp.appendingPathComponent("winstaller-\(UUID().uuidString).err")
        FileManager.default.createFile(atPath: outURL.path, contents: nil)
        FileManager.default.createFile(atPath: errURL.path, contents: nil)
        defer {
            try? FileManager.default.removeItem(at: outURL)
            try? FileManager.default.removeItem(at: errURL)
        }

        let process = Process()
        process.executableURL = URL(fileURLWithPath: command.executable)
        process.arguments = command.arguments
        process.standardOutput = try FileHandle(forWritingTo: outURL)
        process.standardError = try FileHandle(forWritingTo: errURL)

        let startedAt = Date()
        // `process` is confined to this call; terminate() from the cancellation
        // and timeout handlers is safe.
        nonisolated(unsafe) let proc = process

        // Timeout watchdog.
        let timeoutFlag = BoolBox()
        let timeoutTask: Task<Void, Never>? = timeout.interval.map { interval in
            Task {
                try? await Task.sleep(nanoseconds: UInt64(interval * 1_000_000_000))
                if proc.isRunning {
                    timeoutFlag.value = true
                    proc.terminate()
                }
            }
        }

        do {
            try await withTaskCancellationHandler {
                try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
                    proc.terminationHandler = { _ in continuation.resume() }
                    do {
                        try proc.run()
                    } catch {
                        proc.terminationHandler = nil
                        continuation.resume(throwing: CommandRunnerError.launchFailed(error.localizedDescription))
                    }
                }
            } onCancel: {
                if proc.isRunning { proc.terminate() }
            }
        } catch {
            timeoutTask?.cancel()
            throw error
        }
        timeoutTask?.cancel()
        let timedOut = timeoutFlag.value

        let result = CommandResult(
            executable: command.executable,
            arguments: command.arguments,
            standardOutput: (try? Data(contentsOf: outURL)) ?? Data(),
            standardError: (try? Data(contentsOf: errURL)) ?? Data(),
            exitCode: process.terminationStatus,
            startedAt: startedAt,
            finishedAt: Date()
        )

        if Task.isCancelled { throw CommandRunnerError.cancelled }
        if timedOut, let interval = timeout.interval {
            throw CommandRunnerError.timedOut(executable: command.executable, seconds: interval)
        }
        return result
    }
}

/// Records commands and returns canned output. Used by tests and the
/// user-facing "Simulate (dry-run)" mode so no destructive command ever runs.
public final class DryRunCommandRunner: CommandRunning, @unchecked Sendable {
    public struct Stub: Sendable {
        public var standardOutput: Data
        public var standardError: Data
        public var exitCode: Int32

        public init(standardOutput: Data = Data(), standardError: Data = Data(), exitCode: Int32 = 0) {
            self.standardOutput = standardOutput
            self.standardError = standardError
            self.exitCode = exitCode
        }

        public static func text(_ string: String, exitCode: Int32 = 0) -> Stub {
            Stub(standardOutput: Data(string.utf8), exitCode: exitCode)
        }
    }

    private let lock = NSLock()
    private var _invocations: [PlannedCommand] = []
    private let responder: @Sendable (PlannedCommand) -> Stub

    /// - Parameter responder: Supplies canned output per command. Defaults to a
    ///   successful empty result, which is enough to exercise the state machine.
    public init(responder: @escaping @Sendable (PlannedCommand) -> Stub = { _ in Stub() }) {
        self.responder = responder
    }

    public var invocations: [PlannedCommand] {
        lock.withLock { _invocations }
    }

    public func run(_ command: PlannedCommand, timeout: CommandTimeout) async throws -> CommandResult {
        try Task.checkCancellation()
        lock.withLock { _invocations.append(command) }
        let stub = responder(command)
        let now = Date()
        return CommandResult(
            executable: command.executable,
            arguments: command.arguments,
            standardOutput: stub.standardOutput,
            standardError: stub.standardError,
            exitCode: stub.exitCode,
            startedAt: now,
            finishedAt: now
        )
    }
}

File: apps/windows/WInstaller.Core/CommandRunner.cs
using System.Diagnostics;

namespace WInstaller.Core;

/// <summary>
/// The result of running a single command. Command output is treated as
/// untrusted input (TERMINAL_AUTOMATION.md): callers parse it defensively and
/// never evaluate it as code.
/// </summary>
public sealed record CommandResult(
    string Executable,
    IReadOnlyList<string> Arguments,
    string StandardOutput,
    string StandardError,
    int ExitCode,
    DateTimeOffset StartedAt,
    DateTimeOffset FinishedAt)
{
    public bool Succeeded => ExitCode == 0;
}

/// <summary>Timeout classes recommended by TERMINAL_AUTOMATION.md.</summary>
public readonly struct CommandTimeout
{
    public TimeSpan? Interval { get; }

    private CommandTimeout(TimeSpan? interval) => Interval = interval;

    /// <summary>Fast metadata query (generous enough for PowerShell startup).</summary>
    public static readonly CommandTimeout Metadata = new(TimeSpan.FromSeconds(30));
    /// <summary>Mount / unmount an image.</summary>
    public static readonly CommandTimeout Mount = new(TimeSpan.FromSeconds(90));
    /// <summary>Clean + format a removable disk.</summary>
    public static readonly CommandTimeout Format = new(TimeSpan.FromSeconds(300));
    /// <summary>Eject a volume.</summary>
    public static readonly CommandTimeout Eject = new(TimeSpan.FromSeconds(60));
    /// <summary>Long-running work watched by progress rather than a hard timeout.</summary>
    public static readonly CommandTimeout Watched = new(null);

    public static CommandTimeout Seconds(double seconds) => new(TimeSpan.FromSeconds(seconds));
}

public enum CommandRunnerErrorKind
{
    LaunchFailed,
    TimedOut,
    ElevationDeclined,
    Cancelled,
}

public sealed class CommandRunnerException : Exception
{
    public CommandRunnerErrorKind Kind { get; }

    public CommandRunnerException(CommandRunnerErrorKind kind, string message)
        : base(message)
    {
        Kind = kind;
    }
}

/// <summary>
/// Abstraction over command execution so the whole pipeline stays testable
/// without touching a real disk. Commands are always an executable plus an
/// argument array — never a shell-interpolated string.
/// </summary>
public interface ICommandRunner
{
    Task<CommandResult> RunAsync(PlannedCommand command, CommandTimeout timeout, CancellationToken cancellationToken = default);
}

/// <summary>Well-known executable and helper-script locations.</summary>
public static class WindowsCommands
{
    /// <summary>
    /// Absolute path of a tool under System32 (falls back to the bare name off
    /// the Windows system directory, e.g. in unit tests on other OSes).
    /// </summary>
    public static string SystemExecutable(string name)
    {
        var system = Environment.GetFolderPath(Environment.SpecialFolder.System);
        return string.IsNullOrEmpty(system) ? name : Path.Combine(system, name);
    }

    /// <summary>Windows PowerShell 5.1, present on every supported Windows install.</summary>
    public static string PowerShell => SystemExecutable(Path.Combine("WindowsPowerShell", "v1.0", "powershell.exe"));

    /// <summary>Directory holding the bundled .ps1 helpers, next to the app binaries.</summary>
    public static string ScriptsDirectory => Path.Combine(AppContext.BaseDirectory, "Scripts");

    /// <summary>
    /// Builds an argv-only PowerShell invocation of a bundled helper script.
    /// Parameters are passed as separate argv entries, never interpolated into
    /// a command string.
    /// </summary>
    public static PlannedCommand Script(
        string scriptName,
        IReadOnlyList<string> parameters,
        bool isDestructive,
        bool requiresElevation = false)
    {
        var arguments = new List<string>
        {
            "-NoProfile",
            "-NonInteractive",
            "-ExecutionPolicy", "Bypass",
            "-File", Path.Combine(ScriptsDirectory, scriptName),
        };
        arguments.AddRange(parameters);
        return new PlannedCommand(PowerShell, arguments, isDestructive, requiresElevation);
    }
}

/// <summary>
/// Runs real processes. Non-elevated commands are captured via redirected
/// stdout/stderr. Elevated commands (ADR-0003: a single UAC prompt scoped to
/// the destructive step, never the whole app) run through ShellExecute
/// "runas", which cannot redirect output — those scripts report their result
/// through a caller-provided result file instead.
/// </summary>
public sealed class ProcessCommandRunner : ICommandRunner
{
    public async Task<CommandResult> RunAsync(PlannedCommand command, CommandTimeout timeout, CancellationToken cancellationToken = default)
    {
        var startedAt = DateTimeOffset.Now;
        var info = new ProcessStartInfo
        {
            FileName = command.Executable,
            UseShellExecute = command.RequiresElevation,
            CreateNoWindow = true,
            RedirectStandardOutput = !command.RequiresElevation,
            RedirectStandardError = !command.RequiresElevation,
        };
        if (command.RequiresElevation)
        {
            info.Verb = "runas";
            info.WindowStyle = ProcessWindowStyle.Hidden;
        }
        foreach (var argument in command.Arguments)
        {
            info.ArgumentList.Add(argument);
        }

        Process process;
        try
        {
            process = Process.Start(info) ?? throw new CommandRunnerException(
                CommandRunnerErrorKind.LaunchFailed, $"Could not start {command.Executable}.");
        }
        catch (System.ComponentModel.Win32Exception exception) when (exception.NativeErrorCode == 1223)
        {
            // ERROR_CANCELLED: the user declined the UAC prompt.
            throw new CommandRunnerException(
                CommandRunnerErrorKind.ElevationDeclined,
                "The administrator prompt was declined, so the drive was not modified.");
        }
        catch (Exception exception) when (exception is not CommandRunnerException)
        {
            throw new CommandRunnerException(CommandRunnerErrorKind.LaunchFailed, exception.Message);
        }

        using (process)
        {
            var stdoutTask = command.RequiresElevation
                ? Task.FromResult(string.Empty)
                : process.StandardOutput.ReadToEndAsync(cancellationToken);
            var stderrTask = command.RequiresElevation
                ? Task.FromResult(string.Empty)
                : process.StandardError.ReadToEndAsync(cancellationToken);

            using var timeoutSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
            if (timeout.Interval is { } interval)
            {
                timeoutSource.CancelAfter(interval);
            }

            try
            {
                await process.WaitForExitAsync(timeoutSource.Token).ConfigureAwait(false);
            }
            catch (OperationCanceledException)
            {
                try
                {
                    process.Kill(entireProcessTree: true);
                }
                catch
                {
                    // The process may have exited between the timeout and the kill.
                }

                if (cancellationToken.IsCancellationRequested)
                {
                    throw new OperationCanceledException(cancellationToken);
                }
                throw new CommandRunnerException(
                    CommandRunnerErrorKind.TimedOut,
                    $"{Path.GetFileName(command.Executable)} did not finish within {timeout.Interval?.TotalSeconds:0} seconds.");
            }

            var standardOutput = await stdoutTask.ConfigureAwait(false);
            var standardError = await stderrTask.ConfigureAwait(false);

            return new CommandResult(
                command.Executable,
                command.Arguments,
                standardOutput,
                standardError,
                process.ExitCode,
                startedAt,
                DateTimeOffset.Now);
        }
    }
}
